File: src/init_guess_unicycles.cpp
#include "init_guess_unicycles.hpp"
#include <iostream>
#include <fstream>
#include <vector>
#include <yaml-cpp/yaml.h>
#include <Eigen/Dense>
#include <algorithm>


// Pad a matrix to match the maximum number of rows by repeating the last row
Eigen::MatrixXd pad_unicycles_matrix(const Eigen::MatrixXd &matrix, int maxRows) {
    Eigen::MatrixXd padded = Eigen::MatrixXd::Zero(maxRows, matrix.cols());
    int currentRows = matrix.rows();

    // Copy existing rows
    padded.topRows(currentRows) = matrix;

    // Pad remaining rows with the last row
    if (currentRows > 0) {
        for (int i = currentRows; i < maxRows; ++i) {
            padded.row(i) = matrix.row(currentRows - 1);
        }
    }

    return padded;
}

// Save a YAML file
void save_yaml(const std::string &file_path, const YAML::Node &data) {
    std::ofstream outFile(file_path);
    outFile << data;
    outFile.close();
}

void generate_init_guess_unicycles(std::string &envPath,
                        std::string &dbcbsPath, 
                        std::string &resultPath, 
                         size_t numRobots,
                        std::string &joint_robot_env_path) {
    // Load environment YAML
    YAML::Node env = YAML::LoadFile(envPath);

    // Load db_cbs YAML
    YAML::Node dbcbs = YAML::LoadFile(dbcbsPath);

    // Parse robot states and actions
    std::vector<Eigen::MatrixXd> robotStates(numRobots), robotActions(numRobots);
    int maxRowsStates = 0, maxRowsActions = 0;

    for (size_t i = 0; i < numRobots; ++i) {
        auto states = dbcbs["result"][i]["states"];
        auto actions = dbcbs["result"][i]["actions"];

        // Convert YAML to Eigen matrices
        Eigen::MatrixXd stateMatrix(states.size(), states[0].size());
        Eigen::MatrixXd actionMatrix(actions.size(), actions[0].size());

        for (size_t j = 0; j < states.size(); ++j) {
            for (size_t k = 0; k < states[j].size(); ++k) {
                stateMatrix(j, k) = states[j][k].as<double>();
            }
        }

        for (size_t j = 0; j < actions.size(); ++j) {
            for (size_t k = 0; k < actions[j].size(); ++k) {
                actionMatrix(j, k) = actions[j][k].as<double>();
            }
        }

        robotStates[i] = stateMatrix;
        robotActions[i] = actionMatrix;

        maxRowsStates = std::max(maxRowsStates, static_cast<int>(stateMatrix.rows()));
        maxRowsActions = std::max(maxRowsActions, static_cast<int>(actionMatrix.rows()));
    }

    // Pad states and actions to match the maximum size
    std::vector<Eigen::MatrixXd> paddedRobotStates(numRobots), paddedRobotActions(numRobots);
    for (size_t i = 0; i < numRobots; ++i) {
        paddedRobotStates[i] = pad_unicycles_matrix(robotStates[i], maxRowsStates);
        paddedRobotActions[i] = pad_unicycles_matrix(robotActions[i], maxRowsActions);
    }

    // Concatenate actions
    Eigen::MatrixXd concatenatedActions = Eigen::MatrixXd::Zero(maxRowsActions, numRobots * paddedRobotActions[0].cols());
    for (size_t i = 0; i < numRobots; ++i) {
        concatenatedActions.block(0, i * paddedRobotActions[i].cols(), maxRowsActions, paddedRobotActions[i].cols()) = paddedRobotActions[i];
    }

    // Initialize the joint states matrix
    Eigen::MatrixXd unicyclesJointStates = Eigen::MatrixXd::Zero(maxRowsStates, 2 + numRobots + (numRobots - 1));

    // Populate the joint states matrix
    for (size_t i = 0; i < maxRowsStates; ++i) {
        // Add the first robot's state
        unicyclesJointStates(i, 0) = paddedRobotStates[0](i, 0); // px1
        unicyclesJointStates(i, 1) = paddedRobotStates[0](i, 1); // py1
        unicyclesJointStates(i, 2) = paddedRobotStates[0](i, 2); // alpha1

        // Add alpha for remaining robots
        for (size_t j = 1; j < numRobots; ++j) {
            unicyclesJointStates(i, 2 + j) = paddedRobotStates[j](i, 2); // alpha_j
        }

        // Compute theta for each rod
        for (size_t j = 0; j < numRobots - 1; ++j) {
            Eigen::Vector2d pi = paddedRobotStates[j].row(i).head(2);
            Eigen::Vector2d pi_next = paddedRobotStates[j + 1].row(i).head(2);
            Eigen::Vector2d u = pi_next - pi;
            double theta = std::atan2(u.y(), u.x());

            // Add theta to the joint states
            unicyclesJointStates(i, 2 + numRobots + j) = theta;
        }
    }

    // Copy the joint_robot key content into robots
    if (!env["joint_robot"] || env["joint_robot"].size() == 0) {
        throw std::runtime_error("joint_robot key is missing or empty in the provided environment YAML.");
    }

    YAML::Node robotsNode = env["joint_robot"];

    // Create the output environment YAML
    YAML::Node outputEnvYaml = env;
    outputEnvYaml.remove("joint_robot");
    outputEnvYaml["robots"] = robotsNode;

    // Save the new env.yaml file
    std::string envOutputPath = resultPath.substr(0, resultPath.find_last_of("/\\") + 1) + "env.yaml";
    joint_robot_env_path = envOutputPath;
    std::ofstream envOutFile(envOutputPath);
    envOutFile << outputEnvYaml;
    envOutFile.close();


    // Save results to YAML
    YAML::Node result;
    result["result"]["states"] = YAML::Node();
    result["result"]["actions"] = YAML::Node();

    for (int i = 0; i < unicyclesJointStates.rows(); ++i) {
        YAML::Node stateRow;
        for (int j = 0; j < unicyclesJointStates.cols(); ++j) {
            stateRow.push_back(unicyclesJointStates(i, j));
        }
        result["result"]["states"].push_back(stateRow);
    }

    for (int i = 0; i < concatenatedActions.rows(); ++i) {
        YAML::Node actionRow;
        for (int j = 0; j < concatenatedActions.cols(); ++j) {
            actionRow.push_back(concatenatedActions(i, j));
        }
        result["result"]["actions"].push_back(actionRow);
    }

    result["result"]["num_action"] = concatenatedActions.rows();
    result["result"]["num_states"] = unicyclesJointStates.rows();

    save_yaml(resultPath, result);

    std::cout << "Result saved to: " << resultPath << std::endl;


}

File: scripts/init_guess_unicycles.py
import sys
sys.path.append('./')
import numpy as np
import yaml
import argparse
from pathlib import Path
import subprocess


np.set_printoptions(linewidth=np.inf)
np.set_printoptions(suppress=True)

#### This script generates the init guess from the plan of the two single robots to the joint robot
#### moreover, it saves an html file for the meshcat animation of this init guess

def saveyaml(file_dir, data):
    with open(file_dir, 'w') as f:
        yaml.safe_dump(data, f, default_flow_style=None)

def normalize(vec):
    norm_v = np.linalg.norm(vec)
    if norm_v > 0:
        return np.array(vec)/norm_v
    else: 
        raise ValueError("Cannot normalize a zero vector.")

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--env', type=str, help="environment")
    parser.add_argument('--joint_robot_env', default=None, type=str, help="environment")
    parser.add_argument("--num_robots", type=int, help="num of robots")
    parser.add_argument('--payload', default=None, type=str, help="payload pose in dbcbs")
    parser.add_argument('--dbcbs', type=str, help="dbcbs sol")
    parser.add_argument('--result', type=str, help="init_guess_cable.yaml")
    parser.add_argument('--output', type=str, help="init_guess_cable.html")

    args = parser.parse_args()
    num_robots = args.num_robots
    if args.joint_robot_env is None:
        path_to_env = args.env 
    else:
        path_to_env = args.joint_robot_env
    path_to_dbcbs = args.dbcbs
    html_path = args.output
    path_to_result = args.result

    # Load environment
    with open(path_to_env, "r") as f: 
        env = yaml.safe_load(f)

    # Load db_cbs states
    with open(path_to_dbcbs, "r") as f: 
        db_cbs_states = yaml.safe_load(f)

    robots_states = []
    robots_actions = []

    for i in range(num_robots):
        robots_states.append(np.array(db_cbs_states["result"][i]["states"]))
        action_array = np.array(db_cbs_states["result"][i]["actions"])
        robots_actions.append(action_array)

    # Determine the maximum number of rows
    max_rows_states = max(arr.shape[0] for arr in robots_states)
    max_rows_actions = max(arr.shape[0] for arr in robots_actions)

    # Pad the arrays with zeros to match the maximum number of rows
    padded_robot_states = [np.pad(arr, ((0, max_rows_states - arr.shape[0]), (0, 0)), 'edge') for arr in robots_states]
    padded_robot_actions = [np.pad(arr, ((0, max_rows_actions - arr.shape[0]), (0, 0)), 'edge') for arr in robots_actions]
    actions = np.concatenate(padded_robot_actions, axis=1)
    num_states = padded_robot_states[0].shape[0]

    # Initialize joint states with the new representation
    unicycles_joint_states = np.zeros((padded_robot_states[0].shape[0], 2 + num_robots + (num_robots - 1)))

    for i in range(num_states):
        # Add the first robot's state
        unicycles_joint_states[i, 0:3] = padded_robot_states[0][i, 0:3]  # px1, py1, alpha1

        # Add alpha for remaining robots
        for j in range(1, num_robots):
            unicycles_joint_states[i, 2 + j] = padded_robot_states[j][i, 2]  # alpha_j

        # Compute theta for each rod
        for j in range(num_robots - 1):
            pi = padded_robot_states[j][i]
            pi_next = padded_robot_states[j + 1][i]
            u = pi_next[0:2] - pi[0:2] 
            th = np.arctan2(u[1], u[0])
            
            # Add theta to the joint states
            unicycles_joint_states[i, 2 + num_robots + j] = th
    # Save the result to a YAML file
    result_yaml = dict()
    result_yaml["result"] = dict()
    result_yaml["result"]["states"] = unicycles_joint_states.tolist()
    result_yaml["result"]["actions"] = actions.tolist()
    result_yaml["result"]["num_action"] = len(actions.tolist())
    result_yaml["result"]["num_states"] = len(unicycles_joint_states.tolist())

    saveyaml(path_to_result, result_yaml)
    # Optional visualization setup
    script = "../scripts/visualize_unicycles.py"
    subprocess.run(["python3", script, "--robot", "point", "--env", path_to_env, "--result", path_to_result, "--output", html_path])
    print(html_path)
    print(path_to_result)


if __name__ == "__main__":
    main()
